From: Huan Yang <link@vivo.com>
Subject: mm: move the easily assessable conditions forward
Date: Thu, 15 Aug 2024 16:31:01 +0800

Currently try_to_map_unused_to_zeropage() tries to use shared zero page to
save some memory of sub page.

If forbidding zeropage there is no need to do anything rather than
attempting to assess wthether to use it afterwards.

Link: https://lkml.kernel.org/r/20240815083102.653820-1-link@vivo.com
Signed-off-by: Huan Yang <link@vivo.com>
Cc: Yu Zhao <yuzhao@google.com>
Signed-off-by: Andrew Morton <akpm@linux-foundation.org>
---

 mm/migrate.c |    5 ++++-
 1 file changed, 4 insertions(+), 1 deletion(-)

--- a/mm/migrate.c~mm-remap-unused-subpages-to-shared-zeropage-when-splitting-isolated-thp-fix
+++ a/mm/migrate.c
@@ -192,6 +192,9 @@ static bool try_to_map_unused_to_zeropag
 	VM_BUG_ON_PAGE(!PageLocked(page), page);
 	VM_BUG_ON_PAGE(pte_present(*pvmw->pte), page);
 
+	if (mm_forbids_zeropage(pvmw->vma->vm_mm))
+		return false;
+
 	if (PageMlocked(page) || (pvmw->vma->vm_flags & VM_LOCKED))
 		return false;
 
@@ -204,7 +207,7 @@ static bool try_to_map_unused_to_zeropag
 	contains_data = memchr_inv(addr, 0, PAGE_SIZE);
 	kunmap_local(addr);
 
-	if (contains_data || mm_forbids_zeropage(pvmw->vma->vm_mm))
+	if (contains_data)
 		return false;
 
 	newpte = pte_mkspecial(pfn_pte(my_zero_pfn(pvmw->address),
_
